src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp

```diff
@@ -4,161 +4,219 @@
-                // Don't set anything
-                return;
-            }
-        }
-        m_orig_data = tensor.data();
-        m_orig_shape = tensor.get_shape();
-        m_orig_type = tensor.get_element_type();
-    } else if (type == TransformType::CONCAT && std::holds_alternative<ConcatMeta>(transform)) {
-        m_transform = std::make_pair(type, transform);
-    } else if (type == TransformType::UNPACK && std::holds_alternative<UnpackMeta>(transform)) {
-        m_transform = std::make_pair(type, transform);
-    } else {
-        NPUW_ASSERT(false);
-    }
+const ov::Shape& LazyTensorImpl::get_shape() const {
+    return m_orig_shape;
+}
+
+const ov::element::Type& LazyTensorImpl::get_type() const {
+    return m_orig_type;
+}
+
+LazyTensorImpl::LazyTensorImpl(Transform&& t) {
+    std::visit(overloaded{[](const auto& op) {
+                              NPUW_ASSERT(false && "Unsupported LazyTensorImpl constructor!");
+                          },
+                          [this](const op::Tensor& op) {
+                              const auto& tensor = op.tensor;
+                              if (!tensor) {
+                                  // Dummy tensor - don't initialize
+                                  return;
+                              }
+                              m_orig_data = tensor.data();
+                              m_orig_shape = tensor.get_shape();
+                              m_orig_type = tensor.get_element_type();
+                              m_transform = op::Tensor{tensor};
+                          },
+                          [this](const op::Const& op) {
+                              const auto& const_ptr = op.node;
+                              m_orig_data = const_ptr->get_data_ptr();
+                              m_orig_shape = const_ptr->get_shape();
+                              m_orig_type = const_ptr->get_element_type();
+                              m_transform = op::Const{const_ptr};
+                          },
+                          [this](const op::Concat& op) {
+                              m_transform = op;
+                          },
+                          [this](const op::Unpack& op) {
+                              m_transform = op;
+                          },
+                          [this](const op::Permute& op) {
+                              m_transform = op;
+                              m_orig_data = op.tensor.get_data();
+                              m_orig_shape = op.tensor.get_shape();
+                              m_orig_type = op.tensor.get_type();
+                          },
+                          [this](const op::Convert& op) {
+                              m_transform = op;
+                              m_orig_data = op.tensor.get_data();
+                              m_orig_shape = op.tensor.get_shape();
+                              m_orig_type = op.tensor.get_type();
+                          }},
```

Comment:
Not sure what's the reason to have all this processing here. Why it is not simple `: m_transform(std::move(t))`?

Why do you need this `m_orig*` stuff all the way down?

Reply:
Done